action/protocol/poll/staking_command.go

```diff
@@ -114,47 +111,56 @@ func (sc *stakingCommand) Validate(ctx context.Context, act action.Action) error
 }
 
 func (sc *stakingCommand) CalculateCandidatesByHeight(ctx context.Context, height uint64) (state.CandidateList, error) {
-	if sc.stakingV1 == nil {
-		return sc.stakingV2.ActiveCandidates(ctx)
-	}
-
-	if sc.stakingV2 == nil {
-		return sc.stakingV1.CalculateCandidatesByHeight(ctx, height)
-	}
-
 	// transition to V2 starting Fairbank
-	if sc.hu.IsPost(config.Fairbank, height) {
-		return sc.stakingV2.ActiveCandidates(ctx)
+	bcCtx := protocol.MustGetBlockchainCtx(ctx)
+	hu := config.NewHeightUpgrade(&bcCtx.Genesis)
+	if sc.stakingV1 == nil || hu.IsPost(config.Fairbank, height) {
+		cands, err := sc.stakingV2.ActiveCandidates(ctx)
+		if err != nil {
+			return cands, err
+		}
+		bcCtx := protocol.MustGetBlockchainCtx(ctx)
+		return sc.filterAndSortCandidatesByVoteScore(cands, bcCtx.Tip.Timestamp), nil
 	}
 	return sc.stakingV1.CalculateCandidatesByHeight(ctx, height)
 }
 
 // Delegates returns exact number of delegates of current epoch
 func (sc *stakingCommand) Delegates(ctx context.Context, sr protocol.StateReader) (state.CandidateList, error) {
-	// TODO: handle V2
-	return sc.stakingV1.Delegates(ctx, sr)
+	return sc.slasher.GetActiveBlockProducers(ctx, sr, false)
 }
 
 // NextDelegates returns exact number of delegates of next epoch
 func (sc *stakingCommand) NextDelegates(ctx context.Context, sr protocol.StateReader) (state.CandidateList, error) {
-	// TODO: handle V2
-	return sc.stakingV1.Delegates(ctx, sr)
+	return sc.slasher.GetActiveBlockProducers(ctx, sr, true)
 }
 
 // Candidates returns candidate list from state factory of current epoch
 func (sc *stakingCommand) Candidates(ctx context.Context, sr protocol.StateReader) (state.CandidateList, error) {
-	// TODO: handle V2
-	return sc.stakingV1.Candidates(ctx, sr)
+	return sc.slasher.GetCandidates(ctx, sr, false)
 }
 
 // NextCandidates returns candidate list from state factory of next epoch
 func (sc *stakingCommand) NextCandidates(ctx context.Context, sr protocol.StateReader) (state.CandidateList, error) {
-	// TODO: handle V2
-	return sc.stakingV1.NextCandidates(ctx, sr)
+	return sc.slasher.GetCandidates(ctx, sr, true)
 }
 
 func (sc *stakingCommand) ReadState(ctx context.Context, sr protocol.StateReader, method []byte, args ...[]byte) ([]byte, error) {
-	// TODO: handle V2
+	if sc.stakingV1 == nil {
```

Comment:
this is the right order, it returns for stakingV1 == nil
then no need to worry about nil stakingV1 passing this point